src/humancellatlas/data/metadata/api.py

```diff
@@ -42,28 +42,74 @@
 JSON = Dict[str, AnyJSON]
 
 
+@dataclass(init=False)
+class ManifestEntry:
+    json: JSON = field(init=False, repr=False)
+    content_type: str = field(init=False)
+    crc32c: str
+    indexed: bool
+    name: str
+    s3_etag: Optional[str]
+    sha1: Optional[str]
+    sha256: str
+    size: int
+    # only populated if bundle was requested with `directurls` or `directurls` set
+    url: Optional[str]
+    uuid: UUID4 = field(init=False)
+    version: str
+    is_stitched: bool = field(init=False)
+
+    def __init__(self, json: JSON):
+        # '/' was once forbidden in file paths and was encoded with '!'. Now
+        # '/' is allowed and we force it in the metadata so that backwards
+        # compatibility is simplified downstream.
+        json['name'] = json['name'].replace('!', '/')
+        self.json = json
+        self.content_type = json['content-type']
+        self.uuid = UUID4(json['uuid'])
+        self.is_stitched = json.get('is_stitched', False)
+        for f in fields(self):
+            if f.init:
+                value = json.get(f.name)
+                if value is None and not is_optional(f.type):
+                    raise TypeError('Property cannot be absent or None', f.name)
+                else:
+                    setattr(self, f.name, value)
+
+
 @dataclass(init=False)
 class Entity:
     json: JSON = field(repr=False)
     document_id: UUID4
     submitter_id: Optional[str]
+    metadata_manifest_entry: ManifestEntry
+
+    @property
+    def is_stitched(self):
+        return self.metadata_manifest_entry.is_stitched
 
     @classmethod
-    def from_json(cls, json: JSON, **kwargs):
+    def from_json(cls, json: JSON, metadata_manifest_entry: ManifestEntry, **kwargs):
         content = json.get('content', json)
         described_by = content['describedBy']
         schema_name = described_by.rpartition('/')[2]
         try:
             sub_cls = entity_types[schema_name]
         except KeyError:
             raise TypeLookupError(described_by)
-        return sub_cls(json, **kwargs)
+        return sub_cls(json, metadata_manifest_entry, **kwargs)
 
-    def __init__(self, json: JSON) -> None:
+    def __init__(self, json: JSON, metadata_manifest_entry: ManifestEntry) -> None:
         super().__init__()
         self.json = json
+        self.metadata_manifest_entry = metadata_manifest_entry
         provenance = json.get('hca_ingest') or json['provenance']
         self.document_id = UUID4(provenance['document_id'])
+        # Some older DCP/1 bundles use different UUIDs in the manifest and
+        # metadata.
+        if False:
+            # noinspection PyUnreachableCode
```

Comment:
```suggestion
        # noinspection PyUnreachableCode
        if False:
```

Reply:
Thanks, I was struggling to get the syntax right